Prometheus: added time range filter to series labels query

FUSAKLA opened the pull request:
**What this PR does / why we need it**:
This should mitigate query impact on the prometheus datasource and give relevant data for the time range

**Which issue(s) this PR fixes**:
Fixes #16844 

**Special notes for your reviewer**:
Sorry I'm not a front-end type of person and I'm bit of a Alice in Wonderland here :D
But I decided to give it a try.

I tested it and it fills in the end but the start is `NaN` and I have no idea why. 
Could someone please give some hint possibly? Thanks

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/grafana/grafana?pullRequest=16851) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/grafana?pullRequest=16851) before we can accept your contribution.<br/><hr/>**Martin Chodur** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/grafana?pullRequest=16851) it.</sub>

davkal reviewed (changes requested):
`fetchSeriesLabels` saves results in local caches that are used conditionally in other parts. When the date changes those caches will have outdated data. Relatedly, when a user changes the time in the picker, the labels will be outdated. I dont really see a way around invalidating the caches (or removing them completely).

FUSAKLA commented:
Oh, ok this seems bit more complicated than adding 1 row :)
Will there be still need for the cache at all? The query for series labels will be called only once on the time range change and I don't see need to query this again where the cache would help? Or is the cache persisted on reload of the page.

davkal commented:
The cache made more sense when it was not bound to time intervals. I suggest it's OK to remove it.

FUSAKLA commented:
I dig in a bit and found out the issue is different and happens even for already present functionality.

when you call the
https://github.com/grafana/grafana/blob/e4496080ff9d5a21d4fc295090e6f6e0a493e224/public/app/plugins/datasource/prometheus/datasource.ts#L581

then already the `timeSrv.timeRange()` returns the `from` as invalid date.
![image](https://user-images.githubusercontent.com/6112562/57578487-cec8fd00-748d-11e9-9e36-25f3adae2ee3.png)

It happens also for already present functionality for example here
https://github.com/grafana/grafana/blob/e4496080ff9d5a21d4fc295090e6f6e0a493e224/public/app/plugins/datasource/prometheus/completer.ts#L280
from network console:
![image](https://user-images.githubusercontent.com/6112562/57578494-e7391780-748d-11e9-939d-9b6b7159e775.png)

I managed to track it down that if fails to parse the `time.from` here:
https://github.com/grafana/grafana/blob/813e3ffc15442a8199b04671d08b132efe0a55cf/public/app/features/dashboard/services/TimeSrv.ts#L230

It happens for from being set to "6h".
The interesting part is that queries going to Prometheus has the time intervals set OK. 
It happens only for the meta queries for Prometheus.


(I'll be off for a week or so, so I won't have time to look at it further for a while sorry.)

FUSAKLA commented:
I'm back. Found a while to take a look and it looks like the time ranges in TimeSrv are not updated at all for the explore mode.

I checked how it should be updated and found 
https://github.com/grafana/grafana/blob/ceb9f0855bfc5a21d51d22891e6c50f51e4f148a/public/app/features/dashboard/services/TimeSrv.ts#L117

As I said I'm unfamiliar with JS but it looks like the Explore has different params due to the splitting which could cause not loading the time intervals?

If you would find a while @davkal I'd be grateful for any advice or direction on this. 

Btw the cache does not relate to this particular call I wanted to solve that was used just for series names.

FUSAKLA commented:
Hmm now I see that the explore is completely different than the dashboard.
So those language provider utils should in the explore mode get the explore time range and not the data timeSrv range of datasource I guess?

marefr commented:
This is a bit tricky, The best future proof solution would be to provide time range as argument/options to the method since `timeSrv` is an angular service and Explore uses react. However a singleton `timeSrv` is used in explore in a few places, but as you point out the timeSrv aren't receiving updated range. Maybe tricky to remove dependency on `timeSrv` in Prometheus datasource since it may require a lot of changes?

Before we execute a query in Explore we make sure to get the updated range in case time range is specified as relative time:
https://github.com/grafana/grafana/blob/6a76a92afffc05a66a1980b57f111ec237b6d207/public/app/features/explore/state/actions.ts#L402-L411 

If we need to use `timeSrv` in for example `fetchSeriesLabels` we need to decide a good place to update the `timeSrv` in Explore, but I have currently no good idea of where to do this. Need to look more into this.

FUSAKLA commented:
Great, thanks a lot for the input! 

Not sure how to continue with this now. Are you ok with me trying to remove the dependency on `timeSrv` from the Prometheus datasource as you suggested and getting the time range similar way the `runQueries` does as you linked if possible? Did I get it right?

The bigger refactoring is probably out of my league unfortunately regarding my lack of JS skill and knowledge of the code base.

marefr commented:
We have a related PR open (https://github.com/grafana/grafana/pull/17512#issuecomment-500860820) where we need to solve the same problem (I think). Let us think about this some more and see if we can solve our problem first and then we'll come back to this one.

marefr commented:
I should add. If you want to give it a go, yes please try and remove the timeSrv dependency, however that will probably affect all other datasources, I think.

FUSAKLA commented:
Great, thanks for the link. Unfortunately I don't have the time to dig in deeper into this right now so won't be able to give it a go probably I suppose. I'll watch closely on that issue.

FUSAKLA commented:
@marefr I just tried rebasing on the `hugoh/timesrv_poc` and it seems to work as expected, great job :+1: 

So I'll wait till it gets merged and will make it regular PR instead of draft if you are ok with that.

marefr commented:
@FUSAKLA Thanks for the heads up :+1:

marefr commented:
@FUSAKLA we've now merged https://github.com/grafana/grafana/pull/17677

FUSAKLA commented:
Yes, sorry I'm watching the progress unfortunatelly don't have the time now. I'll will take a look next week I hope. 
Thanks for the heads up!

FUSAKLA committed:
fix datasource/prometheus: added time range filter to prom series labels

FUSAKLA commented:
@marefr @davkal 

So I updated the PR, added the time range params to the series query and also added invalidation of the keys/values cache of the language provider.

The cache rounds the start and end to minutes since the `end` in time ranges such as `last 6h` still rises sou it would update all the time.

Is this ok with you? I tested it and it works as expected.

FUSAKLA committed:
fix datasource/prometheus: added missing method to test mock

davkal commented on public/app/plugins/datasource/prometheus/language_provider.ts:
```diff
@@ -332,10 +346,12 @@ export default class PromQlLanguageProvider extends LanguageProvider {
     try {
-      const data = await this.request(`/api/v1/series?match[]=${name}`);
+      const tRange = this.datasource.getTimeRange();
+      const data = await this.request(`/api/v1/series?match[]=${name}&start=${tRange['start']}&end=${tRange['end']}`);
```
I'd use `.` accessors here.

davkal reviewed (approved):
This can go in as is. Thanks for your contribution @FUSAKLA 🎉

FUSAKLA commented:
Oh, that was quick :) thanks!
